air-interpreter/src/marine.rs

```diff
@@ -65,13 +63,21 @@ pub fn invoke_tracing(
     tracing_params: String,
     tracing_output_mode: u8,
 ) -> InterpreterOutcome {
-    if unsafe { !TRACING_IS_INITED } {
-        logger::init_tracing(tracing_params, tracing_output_mode);
-        unsafe {
-            TRACING_IS_INITED = true;
-        }
+    use tracing_subscriber::fmt::format::FmtSpan;
+
+    let builder = tracing_subscriber::fmt()
+        .with_env_filter(tracing_params)
+        .with_span_events(FmtSpan::ENTER | FmtSpan::CLOSE)
+        .with_writer(std::io::stderr);
+
+    if logger::json_output_mode(tracing_output_mode) {
+        let subscriber = builder.json().finish();
+        tracing::subscriber::with_default(subscriber, || execute_air(air, prev_data, data, params, call_results))
+    } else {
+        // Human-readable output.
+        let subscriber = builder.finish();
+        tracing::subscriber::with_default(subscriber, || execute_air(air, prev_data, data, params, call_results))
```

Comment:
This code looks WTF.  I know that it has a reason: `subscriber` has different type in both branches.  It can be simplified with `Box<dyn ...>` or with some feature of `tracing`: `tracing::subscriber::with_default` simply creates `Dispatch` instance inside, and `Dispatch` type is not parametrized by subscriber type.  So:

```rust
    let dispatch = if logger::json_output_mode(tracing_output_mode) {
        let subscriber = builder.json().finish();
        Dispatch::new(subscriber)
    } else {
        // human-readable output
        let subscriber = builder.finish();  // n.b.: this `subscriber` has different type
        Dispatch::new(subscriber)
    };
    tracing::dispatcher::with_default(&dispatch, || execute_air(air, prev_data, data, params, call_results))
```

Reply:
This looks like a leaked abstraction to me. You are trying to use an entity that was wrapped for a reason.

Reply:
Changed though.

Reply:
How do you tell it is a leaking abstraction, but not a more low-level API?

Reply:
There is a section on dispatcher module in the tracing docs so I might be wrong classifying Dispatcher as an internal entity.

Reply:
Anything else to tell about this PR?